src/gc.c

```diff
@@ -90,23 +156,238 @@ static size_t max_collect_interval = 1250000000UL;
+            JL_PRINTF(JL_STDOUT, "\ttype: ");                           \
+            jl_static_show(JL_STDOUT, jl_typeof(obj));                  \
+            JL_PRINTF(JL_STDOUT, "\n");                                 \
+            add_lostval_parent((jl_value_t*)(obj));                     \
+        }                                                               \
+    } while(0);
+
+#else
+#define verify_val(v)
+#define verify_parent(ty,obj,slot,args...)
+#endif
+
+static bigval_t *big_objects = NULL;
+static bigval_t *big_objects_marked = NULL;
+#ifdef OBJPROFILE
+static void *BUFFTY = (void*)0xdeadb00f;
+#endif
+static void *MATY = (void*)0xdeadaa01;
+
+static size_t array_nbytes(jl_array_t*);
+static inline void objprofile_count(void* ty, int old, int sz)
+{
+#ifdef OBJPROFILE
+#ifdef GC_VERIFY
+    if (verifying) return;
+#endif
+    if ((intptr_t)ty <= 0x10)
+        ty = BUFFTY;
+    void **bp = ptrhash_bp(&obj_counts[old], ty);
+    if (*bp == HT_NOTFOUND)
+        *bp = (void*)2;
+    else
+        (*((ptrint_t*)bp))++;
+    bp = ptrhash_bp(&obj_sizes[old], ty);
+    if (*bp == HT_NOTFOUND)
+        *bp = (void*)(1 + sz);
+    else
+        *((ptrint_t*)bp) += sz;
+#endif
+}
+
+static inline void gc_setmark_other(void *o, int mark_mode)
+{
+    _gc_setmark(o, mark_mode);
+    verify_val(o);
+}
+
+#define inc_sat(v,s) v = (v) >= s ? s : (v)+1;
+#define PROMOTE_AGE 1
+
+static inline int gc_setmark_big(void *o, int mark_mode)
+{
+#ifdef GC_VERIFY
+    if (verifying) {
+        _gc_setmark(o, mark_mode);
+        return 0;
+    }
+#endif
+    bigval_t* hdr = bigval_header(o);
+    int bits = gc_bits(o);
+    if (bits == GC_QUEUED || bits == GC_MARKED)
+        mark_mode = GC_MARKED;
+    if ((mark_mode == GC_MARKED) & (bits != GC_MARKED)) {
+        *hdr->prev = hdr->next;
+        if (hdr->next)
+            hdr->next->prev = hdr->prev;
+        hdr->next = big_objects_marked;
+        hdr->prev = &big_objects_marked;
+        if (big_objects_marked)
+            big_objects_marked->prev = &hdr->next;
+        big_objects_marked = hdr;
+    }
+#ifdef OBJPROFILE
+    if (!(bits & GC_MARKED)) {
+        objprofile_count(jl_typeof(o), mark_mode == GC_MARKED, hdr->sz);
+    }
+#endif
+    _gc_setmark(o, mark_mode);
+    verify_val(o);
+    return mark_mode;
+}
+
+static inline int gc_setmark_pool(void *o, int mark_mode)
+{
+#ifdef GC_VERIFY
+    if (verifying) {
+        _gc_setmark(o, mark_mode);
+        return mark_mode;
+    }
+#endif
+    gcpage_t* page = GC_PAGE(o);
+    int bits = gc_bits(o);
+    if (bits == GC_QUEUED || bits == GC_MARKED) {
+        mark_mode = GC_MARKED;
+    }
+#ifdef OBJPROFILE
+    if (!(bits & GC_MARKED)) {
+        objprofile_count(jl_typeof(o), mark_mode == GC_MARKED, page->osize);
+    }
+#endif
+    _gc_setmark(o, mark_mode);
+    page->gc_bits |= mark_mode;
+    verify_val(o);
+    return mark_mode;
+}
+
+
+static inline int gc_setmark(void *o, int sz, int mark_mode)
```

Comment:
in some cases, I think this is being called with a `sz` value that doesn't include the `jl_value_t *type` tag (e.g. when it comes from `jl_datatype_size`)